tests/e2e/ppnd-e2e/ppnd_test.go

```diff
@@ -0,0 +1,372 @@
+			}, true, false)
+
+			verifyPipelineIsSlowToPause()
+			allowDataLoss()
+			DeletePipelineRollout(slowPipelineRolloutName)
+
+		}
+	})
+
+	It("Should allow data loss in the Pipeline if requested - ISBService update", func() {
+		if UpgradeStrategy == config.PPNDStrategyID {
+
+			createSlowPipelineRollout()
+
+			Document("Updating ISBService to cause a PPND change")
+			updatedISBServiceSpec := isbServiceSpec
+			updatedISBServiceSpec.JetStream.Version = updatedJetstreamVersion
+			rawSpec, err := json.Marshal(updatedISBServiceSpec)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// we use the direct k8s function since we need to control the PipelineRollout to complete the update
+			UpdateISBServiceRolloutInK8S(isbServiceRolloutName, func(rollout apiv1.ISBServiceRollout) (apiv1.ISBServiceRollout, error) {
+				rollout.Spec.InterStepBufferService.Spec.Raw = rawSpec
+				return rollout, nil
+			})
+
+			verifyPipelineIsSlowToPause()
+			allowDataLoss()
+			DeletePipelineRollout(slowPipelineRolloutName)
+
+			// confirm update
+			VerifyISBServiceSpec(Namespace, isbServiceRolloutName, func(retrievedISBServiceSpec numaflowv1.InterStepBufferServiceSpec) bool {
+				return retrievedISBServiceSpec.JetStream.Version == updatedJetstreamVersion
+			})
+
+		}
+	})
+
+	It("Should allow data loss in the Pipeline if requested - Numaflow Controller update", func() {
+		if UpgradeStrategy == config.PPNDStrategyID {
+
+			createSlowPipelineRollout()
+
+			Document("Updating Numaflow controller to cause a PPND change")
+			updatedNumaflowControllerROSpec := apiv1.NumaflowControllerRolloutSpec{
+				Controller: apiv1.Controller{Version: updatedNumaflowControllerVersion},
+			}
+			UpdateNumaflowControllerRolloutInK8S(func(rollout apiv1.NumaflowControllerRollout) (apiv1.NumaflowControllerRollout, error) {
+				rollout.Spec = updatedNumaflowControllerROSpec
+				return rollout, nil
+			})
+
+			verifyPipelineIsSlowToPause()
+			allowDataLoss()
+			DeletePipelineRollout(slowPipelineRolloutName)
+
+			// confirm update
+			VerifyNumaflowControllerDeployment(Namespace, func(d appsv1.Deployment) bool {
+				colon := strings.Index(d.Spec.Template.Spec.Containers[0].Image, ":")
+				return colon != -1 && d.Spec.Template.Spec.Containers[0].Image[colon+1:] == "v"+updatedNumaflowControllerVersion
+			})
+
+		}
+	})
+
+	It("Should update a Pipeline even if the Pipeline is failed", func() {
+
+		// add bad edge to automatically fail Pipeline
+		failedPipelineSpec := initialPipelineSpec
+		failedPipelineSpec.Edges = append(failedPipelineSpec.Edges, numaflowv1.Edge{From: "not", To: "valid"})
+
+		CreatePipelineRollout(failedPipelineRolloutName, Namespace, failedPipelineSpec, true)
+		VerifyPipelineFailed(Namespace, failedPipelineRolloutName)
+
+		time.Sleep(5 * time.Second)
+
+		// update spec to have topology change
+		failedPipelineSpec = updatedPipelineSpec
+
+		// need to update
+		// this update technically doesnt cause data loss since pipeline for now is failed
```

Comment:
regarding this comment, I think if the `dataLoss` field basically means "the field(s) being changed are fields that result in data loss", then it's good to set this true. Perhaps the comment in the function for what the field means, or possibly even the field, could say that?

Reply:
then you can remove this TODO

Reply:
Ack, I think the comment originally had to do with me making a few changes to the update function to handle a failed pipeline and the no data loss update followed the logic we needed for that. We should be good now to remove the comment.